File: examples/ch04/ts/pricing.ts
// Pricing — answers to one actor: finance/commerce rules. One reason to change:
// the money math. No I/O, no email, no disk — which is exactly why it is testable
// in isolation.

import { type Order, subtotal } from "./models";

const TAX_RATES: Record<string, number> = { US: 0.07, DE: 0.19, JP: 0.1 };
const EU_COUNTRIES = ["DE", "FR", "NL"];

export const memberDiscount = (sub: number, isMember: boolean): number =>
  isMember ? sub * 0.85 : sub;

export const taxFor = (amount: number, country: string): number =>
  amount * (TAX_RATES[country] ?? 0);

export const shippingFor = (country: string, giftWrap: boolean): number => {
  let base: number;
  if (country === "US") {
    base = 5.0;
  } else if (EU_COUNTRIES.includes(country)) {
    base = 9.9;
  } else {
    base = 24.9;
  }
  return base + (giftWrap ? 3.5 : 0);
};

export const orderTotal = (order: Order): number => {
  const discounted = memberDiscount(subtotal(order), order.customer.isMember);
  const tax = taxFor(discounted, order.country);
  const shipping = shippingFor(order.country, order.giftWrap);
  return Math.round((discounted + tax + shipping) * 100) / 100;
};

File: examples/ch04/pricing.py
"""Pricing — answers to one actor: finance/commerce rules.

One reason to change: the money math. No I/O, no email, no disk — which is
exactly why it is testable in isolation.
"""

from models import Order

TAX_RATES = {"US": 0.07, "DE": 0.19, "JP": 0.10}
EU_COUNTRIES = ("DE", "FR", "NL")


def member_discount(subtotal: float, is_member: bool) -> float:
    return subtotal * 0.85 if is_member else subtotal


def tax_for(amount: float, country: str) -> float:
    return amount * TAX_RATES.get(country, 0.0)


def shipping_for(country: str, gift_wrap: bool) -> float:
    if country == "US":
        base = 5.00
    elif country in EU_COUNTRIES:
        base = 9.90
    else:
        base = 24.90
    return base + (3.50 if gift_wrap else 0.0)


def order_total(order: Order) -> float:
    discounted = member_discount(order.subtotal, order.customer.is_member)
    tax = tax_for(discounted, order.country)
    shipping = shipping_for(order.country, order.gift_wrap)
    return round(discounted + tax + shipping, 2)
